pkg/generic/thrift/http.go

```diff
@@ -18,71 +18,109 @@ package thrift
+// SetWriteHTTPRequest ...
+func (w *WriteHTTPRequest) SetWriteHTTPRequest(opts *conv.Options, method string) {
+	w.dyOpts = opts
+	w.method = method
 }
 
 // ReadHTTPResponse implement of MessageReaderWithMethod
 type ReadHTTPResponse struct {
-	svc          *descriptor.ServiceDescriptor
-	base64Binary bool
+	svc             *descriptor.ServiceDescriptor
+	enableDynamicgo bool
+	dyOpts          *conv.Options
+	msg             remote.Message
 }
 
 var _ MessageReader = (*ReadHTTPResponse)(nil)
 
 // NewReadHTTPResponse ...
 // Base64 encoding for binary is enabled by default.
 func NewReadHTTPResponse(svc *descriptor.ServiceDescriptor) *ReadHTTPResponse {
-	return &ReadHTTPResponse{svc, true}
+	return &ReadHTTPResponse{svc, false, &conv.Options{}, nil}
 }
 
-// SetBase64Binary enable/disable Base64 encoding for binary.
-// Note that this method is not concurrent-safe.
-func (r *ReadHTTPResponse) SetBase64Binary(enable bool) {
-	r.base64Binary = enable
+// SetReadHTTPResponse ...
+func (r *ReadHTTPResponse) SetReadHTTPResponse(enable bool, opts *conv.Options, msg remote.Message) {
+	r.enableDynamicgo = enable
+	r.dyOpts = opts
+	r.msg = msg
 }
 
 // Read ...
 func (r *ReadHTTPResponse) Read(ctx context.Context, method string, in thrift.TProtocol) (interface{}, error) {
-	fnDsc, err := r.svc.LookupFunctionByMethod(method)
+	// Transport protocol should be TTHeader, Framed, or TTHeaderFramed to enable dynamicgo
+	if !r.enableDynamicgo || r.msg.PayloadLen() == 0 {
+		fnDsc, err := r.svc.LookupFunctionByMethod(method)
+		if err != nil {
+			return nil, err
+		}
+		fDsc := fnDsc.Response
+		return skipStructReader(ctx, in, fDsc, &readerOption{forJSON: true, http: true, binaryWithBase64: !r.dyOpts.NoBase64Binary})
+	}
+
+	// dynamicgo logic
+	tProt, ok := in.(*cthrift.BinaryProtocol)
+	if !ok {
+		return nil, perrors.NewProtocolErrorWithMsg("TProtocol should be BinaryProtocol")
+	}
+
+	if r.svc.DynamicgoDsc == nil {
+		return nil, perrors.NewProtocolErrorWithMsg("svcDsc.DynamicgoDsc is nil")
+	}
+	tyDsc := r.svc.DynamicgoDsc.Functions()[method].Response()
+
+	msgBeginLen := bthrift.Binary.MessageBeginLength(method, thrift.TMessageType(r.msg.MessageType()), r.msg.RPCInfo().Invocation().SeqID())
+	transBuf, err := tProt.ByteBuffer().ReadBinary(r.msg.PayloadLen() - msgBeginLen - bthrift.Binary.MessageEndLength())
 	if err != nil {
 		return nil, err
 	}
-	fDsc := fnDsc.Response
-	return skipStructReader(ctx, in, fDsc, &readerOption{forJSON: true, http: true, binaryWithBase64: r.base64Binary})
+
+	buf := make([]byte, 0, len(transBuf)*2)
+	resp := descriptor.NewHTTPResponse()
+	ctx = context.WithValue(ctx, conv.CtxKeyHTTPResponse, resp)
+	cv := t2j.NewBinaryConv(*r.dyOpts)
+	// decode with dynamicgo
+	// thrift []byte to json []byte
+	if err = cv.DoInto(ctx, tyDsc, transBuf, &buf); err != nil {
+		return nil, err
+	}
+
+	// unwrap one level of json string
+	if len(buf) > structWrapLen {
```

Comment:
You can use a loop to unnest the top level the same as the Write function here, but first you should make sure that dynamicgo supports parsing binary with complete headers but carrying tail data

Reply:
fixed